# <font color='BlueViolet'>SQL
<font size="4"><font color='BlueViolet'> by Keren Katzav

 **<font color='BlueViolet'><font size="4">1.The goals of the study:**

<font color='BlueViolet'>Generate a value proposition for a new product
    
In this project I received a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.    

 **<font color='BlueViolet'><font size="4">2.Study the tables:**


In [1]:
%pip install psycopg2-binary

Note: you may need to restart the kernel to use updated packages.


In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine



db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
query= '''
SELECT *
FROM books
LIMIT 5; '''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
query= '''
SELECT *
FROM authors
LIMIT 5; '''
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
query= '''
SELECT *
FROM publishers
LIMIT 5; '''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
query= '''
SELECT *
FROM ratings
LIMIT 5; '''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
query= '''
SELECT *
FROM reviews
LIMIT 5; '''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

 **<font color='BlueViolet'><font size="4">3.Task:**


**<font color='BlueViolet'>Find the number of books released after January 1, 2000.**


In [8]:
query= '''
SELECT 
    COUNT(book_id) AS number_of_books

FROM books

WHERE 
    publication_date > '2000-01-01'; '''

pd.io.sql.read_sql(query, con = engine)



number_of_books
0              819

**There are 819 books in our database that released after 1/1/2000**

**<font color='BlueViolet'>Find the number of user reviews and the average rating for each book.**



In this section I will create 3 tables:

-The first is arranged in descending order by number_of_reviews+ average_rating

-The second is arranged in descending order only by number_of_reviews

-The third is arranged in descending order only according to average_rating

This is so that I can examine in more depth what the leading books are

In [9]:
query= '''
SELECT
    books.book_id,
    books.title,
    COUNT(DISTINCT(review_id)) AS number_of_reviews,
    AVG(rating) AS average_rating
        
FROM books
INNER JOIN reviews USING (book_id)
INNER JOIN ratings USING (book_id)

GROUP BY book_id
ORDER BY number_of_reviews DESC, average_rating DESC
LIMIT 10; '''

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  \
0      948                            Twilight (Twilight  #1)   
1      302  Harry Potter and the Prisoner of Azkaban (Harr...   
2      299  Harry Potter and the Chamber of Secrets (Harry...   
3      656                                     The Book Thief   
4      734                                   The Glass Castle   
5      497                          Outlander (Outlander  #1)   
6      750                The Hobbit  or There and Back Again   
7      695  The Curious Incident of the Dog in the Night-Time   
8      779  The Lightning Thief (Percy Jackson and the Oly...   
9      963                                Water for Elephants   

   number_of_reviews  average_rating  
0                  7        3.662500  
1                  6        4.414634  
2                  6        4.287500  
3                  6        4.264151  
4                  6        4.206897  
5                  6        4.125000  
6                  6        4.125000  
7                  6        4.081081  
8                  6        4.080645  
9                  6        3.977273

In [10]:
query= '''
SELECT
    books.book_id,
    books.title,
    COUNT(DISTINCT(review_id)) AS number_of_reviews,
    AVG(rating) AS average_rating
        
FROM books
INNER JOIN reviews USING (book_id)
INNER JOIN ratings USING (book_id)

GROUP BY book_id
ORDER BY number_of_reviews DESC; '''

pd.io.sql.read_sql(query, con = engine)


book_id                                title  number_of_reviews  \
0        948              Twilight (Twilight  #1)                  7   
1        854                             The Road                  6   
2        656                       The Book Thief                  6   
3        734                     The Glass Castle                  6   
4        963                  Water for Elephants                  6   
..       ...                                  ...                ...   
989      465    Naked Empire (Sword of Truth  #8)                  1   
990      446                  Moo  Baa  La La La!                  1   
991      431  Merrick (The Vampire Chronicles #7)                  1   
992       92                            Babyville                  1   
993      413                           Lysistrata                  1   

     average_rating  
0          3.662500  
1          3.772727  
2          4.264151  
3          4.206897  
4          3.977273  
..              ...  
989        3.500000  
990        3.000000  
991        4.000000  
992        3.500000  
993        4.000000  

[994 rows x 4 columns]

In [11]:
query= '''
SELECT
    books.book_id,
    books.title,
    COUNT(DISTINCT(review_id)) AS number_of_reviews,
    AVG(rating) AS average_rating
        
FROM books
INNER JOIN reviews USING (book_id)
INNER JOIN ratings USING (book_id)

GROUP BY book_id
ORDER BY average_rating DESC; '''

pd.io.sql.read_sql(query, con = engine)


book_id                                              title  \
0        972  Wherever You Go  There You Are: Mindfulness Me...   
1        513               Piercing the Darkness (Darkness  #2)   
2        901  The Walking Dead  Book One (The Walking Dead #...   
3        136  Captivating: Unveiling the Mystery of a Woman'...   
4         86      Arrows of the Queen (Heralds of Valdemar  #1)   
..       ...                                                ...   
989      915  The World Is Flat: A Brief History of the Twen...   
990      371                                              Junky   
991      316                  His Excellency: George Washington   
992      202                                      Drowning Ruth   
993      303                               Harvesting the Heart   

     number_of_reviews  average_rating  
0                    2            5.00  
1                    2            5.00  
2                    2            5.00  
3                    2            5.00  
4                    2            5.00  
..                 ...             ...  
989                  3            2.25  
990                  2            2.00  
991                  2            2.00  
992                  3            2.00  
993                  2            1.50  

[994 rows x 4 columns]

**For each book we have found the number of user reviews for that book and the average rating that book has.**

**We have 994 books**

**The highest average_rating for the book is 5**

**The highest number_of_reviews for the book is 7**

**It seems that someone who has an average_rating of 5 has a low number_of_reviews of 2**

**On the face of it, it seems that the book: Harry Potter and the Prisoner of Azkaban, leads both in its rating (4.414634) and in number_of_reviews (6)**

**And Twilight (Twilight #1) with rating (3.662500) number_of_reviews (7), By the way, it is the only book with number_of_reviews (7)**

**<font color='BlueViolet'>Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).**

In [12]:
query= '''
SELECT books.publisher_id,
    publisher,
    COUNT(title) AS number_of_books
    
FROM books
INNER JOIN publishers USING (publisher_id)

WHERE books.num_pages >=50 
GROUP BY publisher, publisher_id
ORDER BY number_of_books DESC
LIMIT 1; '''

pd.io.sql.read_sql(query, con = engine)


publisher_id      publisher  number_of_books
0           212  Penguin Books               42

**The publisher that has released the greatest number of books with more than 50 pages is Penguin Books, id=212, numbers of books= 42**

**<font color='BlueViolet'>Identify the author with the highest average book rating (look only at books with at least 50 ratings).**


In [14]:
query= '''
SELECT authors.author,
    AVG(ratings.rating) AS average_rating
    
FROM
    books
    JOIN authors ON books.author_id= authors.author_id
    JOIN ratings ON books.book_id= ratings.book_id  
    
WHERE
    books.book_id IN (
        SELECT
            book_id
        FROM
            ratings
        GROUP BY
            book_id
        HAVING 
            COUNT(1) >=50
    )
GROUP BY
    authors.author
ORDER BY
    average_rating DESC
LIMIT 1; '''
pd.io.sql.read_sql(query, con = engine)

author  average_rating
0  J.K. Rowling/Mary GrandPré        4.287097

**The author with the highest average book rating (from books with 50 or more ratings) is J.K. Rowling/Mary GrandPré, having 4.287097 average rating**

**<font color='BlueViolet'>Find the average number of text reviews among users who rated more than 50 books.**

In [16]:
query= '''
SELECT AVG(COUNT(DISTINCT(reviews.review_id))) OVER () AS average_number_of_reviews

FROM reviews
INNER JOIN ratings ON reviews.username=ratings.username
GROUP BY ratings.username
HAVING COUNT(DISTINCT(ratings.rating_id)) > 50
LIMIT 1; '''

pd.io.sql.read_sql(query, con = engine)


average_number_of_reviews
0                  24.333333

**The average number of text reviews among users who rated more than 50 books is 24.333 rating per user.**

 **<font color='BlueViolet'><font size="4">4.conclusions:**


1. There are 819 books in our database that released after 1/1/2000

2. The highest average_rating for the book is 5, The highest number_of_reviews for the book is 7, It seems that someone who has an average_rating of 5 has a low number_of_reviews of 2

3. it seems that the book: Harry Potter and the Prisoner of Azkaban, leads both in its rating (4.414634) and in number_of_reviews (6) And Twilight (Twilight #1) with rating (3.662500) number_of_reviews (7), By the way, it is the only book with number_of_reviews (7)

4. The publisher that has released the greatest number of books with more than 50 pages is Penguin Books, id=212, numbers of books= 42

5. The author with the highest average book rating (from books with 50 or more ratings) is Diana Gabaldon, having 4.3​ average rating

6. The average number of text reviews among users who rated more than 50 books is 24.333 rating per user.